<table style="float:left">
    <tr>
        <td>
            <img src="emlyon.png" style="height:60px; float:left; padding-right:10px; margin-top:5px" />
        </td>
        <td style="padding-bottom:10px; text-align:left;">
            <h1 style="border-bottom: 1px solid #eeeeee;"> Python BootCamp </h1><br/>
            <span style="display:inline-block; margin-top:-15px;">
            <a href="https://masters.em-lyon.com/en/msc-in-digital-marketing-data-science">[Emlyon]</a> MSc in Digital Marketing & Data Science (DMDS)    
            <br/>
            Sep 2024, Paris | © Saeed VARASTEH [RP] | Jean SAVINIEN
            </span>
        </td>
    </tr>
</table>

<div class="alert-info" style="border-bottom: solid 1px lightgray; background-color:#fff4e3;">
    <img src="homework.png" style="height:60px; float:left; padding-right:10px;" />
    <span style="font-weight:bold; color:#db9425">
        <h4 style="padding-top:25px;"> HOMEWORK 05 </h4>
    </span>
</div>

In [3]:
Name = ""

### Homework 05 - Beauty Products

The aim of this exercise is to assess your ability to apply your newly acquired coding skills to a __practical/real-world__ project.

Our goal is __data exploration__, not modeling.

---

<div style="color:gray; background-color:#eeeeee">
    Dataset description
</div>
<br/>

The __beautyproducts.csv__ file contains information on beauty products sold in more than thirty thousand Amazon stores over the past 23 years. It includes the following features for each product:

- __title__: Name of the product.
- __average_rating__: Rating of the product shown on the product page.
- __rating_number__: Number of ratings in the product.
- __features__: Features of the product.
- __description__: Description of the product.
- __images__: Images of the product. Each image has different sizes (thumb, large, hi_res).
- __store__: Store name of the product.
- __details__: Product details, including materials, brand, sizes, etc.

Lets read the file into a Pandas DataFrame to get a first impression.

In [1]:
# libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Pandas Options
pd.set_option("display.max_rows", 50)

In [3]:
df = pd.read_csv('beautyproducts.csv')
print(df.shape)
df.head()

(112590, 8)


title  average_rating  \
0  Howard LC0008 Leather Conditioner, 8-Ounce (4-...             4.8   
1  Yes to Tomatoes Detoxifying Charcoal Cleanser ...             4.5   
2   Eye Patch Black Adult with Tie Band (6 Per Pack)             4.4   
3  Tattoo Eyebrow Stickers, Waterproof Eyebrow, 4...             3.1   
4  Precision Plunger Bars for Cartridge Grips – 9...             4.3   

   rating_number                                           features  \
0             10                                                 []   
1              3                                                 []   
2             26                                                 []   
3            102                                                 []   
4              7  ['Material: 304 Stainless Steel; Brass tip', '...   

                                         description  \
0                                                 []   
1                                                 []   
2                                                 []   
3                                                 []   
4  ['The Precision Plunger Bars are designed to w...   

                                              images                   store  \
0  [{'thumb': 'https://m.media-amazon.com/images/...         Howard Products   
1  [{'thumb': 'https://m.media-amazon.com/images/...                  Yes To   
2  [{'thumb': 'https://m.media-amazon.com/images/...  Levine Health Products   
3  [{'thumb': 'https://m.media-amazon.com/images/...                Cherioll   
4  [{'thumb': 'https://m.media-amazon.com/images/...               Precision   

                                             details  
0  {'Package Dimensions': '7.1 x 5.5 x 3 inches; ...  
1  {'Item Form': 'Powder', 'Skin Type': 'Acne Pro...  
2         {'Manufacturer': 'Levine Health Products'}  
3  {'Brand': 'Cherioll', 'Item Form': 'Powder', '...  
4                            {'UPC': '644287689178'}

You need to know that working with real-world data is not as straightforward as it was in the classroom with mostly synthetic/clean data sets. 

Most of the time you will be working with large datasets like this one. Or you will be dealing with columns such as __features__, __description__, __images__, and __details__ that contain multiple values as a list or dictionary, which will add another layer of data processing as we work with these columns. 

For example, let us select a product from the "Estee Lauder" store and see what we have on it. e.g. index 4249.

In [8]:
idx = 4249
df.iloc[idx]

title             Bronze Goddess Summer Glow Lip & Cheek Color 0...
average_rating                                                  4.7
rating_number                                                    12
features          ['Bronze Goddess Summer Glow Lip & Cheek Color...
description       ['Estee Lauder Bronze Goddess Summer Glow Lip ...
images            [{'thumb': 'https://m.media-amazon.com/images/...
store                                                  Estee Lauder
details           {'Brand': 'Estee Lauder', 'Color': 'Bronze', '...
Name: 4249, dtype: object

In [10]:
# title
df.iloc[idx].title

'Bronze Goddess Summer Glow Lip & Cheek Color 03 Sunburst'

Titles are stored as __text__ exactly as we can find them on the Amazon webpage.

In [13]:
# average_rating and rating_number
df.iloc[idx].average_rating, df.iloc[idx].rating_number

(4.7, 12)

The average rating and rating number columns have __numeric__ values.

In [16]:
# features
df.iloc[idx].features

"['Bronze Goddess Summer Glow Lip & Cheek Color 03 Sunburst', 'Estee Lauder', 'Full Size', 'Unboxed']"

Features should be stored as a __list of strings__. But if you look closely, the output is also quoted. That means the type is string. We need to fix this because we want the output to be a list so that we can select an element from it.

<span style="color:crimson">Note:</span> This problem is common in real-world data sets.

To convert these strings into actual Python lists, you can use `ast.literal_eval()` from the `ast` module, which safely evaluates strings containing Python expressions and converts them to the corresponding data type. Here's how you can apply it:

In [19]:
import ast
ast.literal_eval( df.iloc[idx].features )

['Bronze Goddess Summer Glow Lip & Cheek Color 03 Sunburst',
 'Estee Lauder',
 'Full Size',
 'Unboxed']

Good! You can simply apply this function to the entire column to convert the values from strings (objects) to lists.

In [22]:
df["features"] = df["features"].apply(ast.literal_eval)

Let us check our previous code again:

In [25]:
# features
df.iloc[idx].features

['Bronze Goddess Summer Glow Lip & Cheek Color 03 Sunburst',
 'Estee Lauder',
 'Full Size',
 'Unboxed']

Voila! Lets do this for __description__, __images__ and __details__ columns as well.

In [28]:
df["description"] = df["description"].apply(ast.literal_eval)
df["images"] = df["images"].apply(ast.literal_eval)
df["details"] = df["details"].apply(ast.literal_eval)

In [29]:
# description
df.iloc[idx].description

['Estee Lauder Bronze Goddess Summer Glow Lip & Cheek Color 03 Sunburst Full Size Unboxed']

In [30]:
# images
df.iloc[idx].images

[{'thumb': 'https://m.media-amazon.com/images/I/31fgVZj+6LL._SS40_.jpg',
  'large': 'https://m.media-amazon.com/images/I/31fgVZj+6LL.jpg',
  'variant': 'MAIN',
  'hi_res': 'https://m.media-amazon.com/images/I/51D9GvPsw5L._SL1399_.jpg'},
 {'thumb': 'https://m.media-amazon.com/images/I/31fgVZj+6LL._SS40_.jpg',
  'large': 'https://m.media-amazon.com/images/I/31fgVZj+6LL.jpg',
  'variant': 'PT01',
  'hi_res': 'https://m.media-amazon.com/images/I/51D9GvPsw5L._SL1399_.jpg'},
 {'thumb': 'https://m.media-amazon.com/images/I/31fgVZj+6LL._SS40_.jpg',
  'large': 'https://m.media-amazon.com/images/I/31fgVZj+6LL.jpg',
  'variant': 'PT02',
  'hi_res': 'https://m.media-amazon.com/images/I/51D9GvPsw5L._SL1399_.jpg'}]

Image information is stored as a __list of dictionaries__. For each image we have four features (thumb, large, variant, hi_res) in a dictionary-style format. For example, to get the high resolution version of the first image, we can write:

In [35]:
df.iloc[idx].images[0]["hi_res"]

'https://m.media-amazon.com/images/I/51D9GvPsw5L._SL1399_.jpg'

Click the link to see how reliable the links are in this dataset.

In [38]:
# details
df.iloc[idx].details

{'Brand': 'Estee Lauder',
 'Color': 'Bronze',
 'Skin Type': 'All',
 'Item Form': 'Cream',
 'Age Range (Description)': 'Adult',
 'Is Discontinued By Manufacturer': 'No',
 'Product Dimensions': '1 x 1 x 5 inches; 4.96 Ounces',
 'UPC': '887167219793'}

The data in this column is stored as a __dictionary__. 

This column stores many details about the product and it is different from one product to another! change the __idx__ to see for yourself.

---

<div style="color:gray; background-color:#eeeeee">
    Your Task
</div>

You are a data analyst in a startup company targeting the beauty market. You are not given a specific task to do! Your job is to extract as much information as you can from this dataset to help your company's marketing strategies.

Here are some questions you can answer.

- How many stores are in the dataset? 
- Which stores sell the most number of items?
- How many brands are selling items in the dataset?
- Which brands have the highest ratings? 
- Which products have the highest average rating? List the top 50. Use products with number of rating greater than 1000.
- Do these top-rated products share common characteristics?
- Display images and print details of the top 10 highest-rated products.
- What will be a good marketing strategy when your company launches its first product?

Add at least __7__ more questions of your choice to this list and try to answer them.

Comment/document everything.

## How many stores are in the dataset?


In [41]:
dff = df.dropna()

In [43]:
dff['store'].value_counts() #count the values in column store (each value is a store)

store
Bath & Body Works    481
WILLBOND             388
Hairro               381
Generic              369
S-noilite            340
                    ... 
ASSOS Cycling          1
LA Bodies              1
Lingon                 1
STAUB                  1
DVIO                   1
Name: count, Length: 30759, dtype: int64

In [45]:
len(dff["store"].unique()) #unique() includes NaN values by default

30759

In [47]:
dff["store"].nunique() #nunique() ignores NaN values by default

30759

## Which stores sell the most number of items?

Because Bath & Body Works is the store that appears the most in the dataframe, we can say that it is the store that sell the msot number of items, given that beautyproducts.csv file contains information on beauty products sold. So each time that a company appears, it means a sold item by them. 

In [113]:
df['store'].value_counts() #counts the occurrences of each value in the store column 

store
Bath & Body Works    481
WILLBOND             388
Hairro               381
Generic              369
S-noilite            340
                    ... 
STAUB                  1
YUATCYA                1
starbuzz               1
Heladerm               1
DVIO                   1
Name: count, Length: 30761, dtype: int64

## How many brands are selling items in the dataset?

To see "Brand" in "details" would be a way to count how many brands are in the df. However, given the lack of information in some of these variables, it can be assumed that a store would be the equivalent to a brand. 

In [115]:
df['store'].value_counts()

store
Bath & Body Works    481
WILLBOND             388
Hairro               381
Generic              369
S-noilite            340
                    ... 
STAUB                  1
YUATCYA                1
starbuzz               1
Heladerm               1
DVIO                   1
Name: count, Length: 30761, dtype: int64

In [117]:
#has no brand in details  
df['details'][4]

{'UPC': '644287689178'}

In [119]:
#has a brand 
df['details'][1]

{'Item Form': 'Powder',
 'Skin Type': 'Acne Prone',
 'Brand': 'Yes To',
 'Age Range (Description)': 'Adult',
 'Unit Count': '10 Fl Oz',
 'Is Discontinued By Manufacturer': 'No',
 'Item model number': 'SG_B076WQZGPM_US',
 'UPC': '653801351125',
 'Manufacturer': 'Yes to Tomatoes'}

If we try to get the brands in the details column we can do this:

In [41]:
#source https://stackoverflow.com/questions/25231989/how-to-check-if-a-variable-is-a-dictionary-in-python
from typing import Dict

In [43]:
#function to get the key and value of the dictionary where the key is brand
def branded(x):
    if isinstance(x, Dict): #need to check if it is a dictionary because if it is not it won't run and this makes sure the iteration goes to all keys 
        for key, value in x.items():
            if key == 'Brand':
                return value
    else:
        return None

In [45]:
df["brand"] = df["details"].apply(branded) #apply function to df and create column brand 
df["brand"]

0               None
1             Yes To
2               None
3           Cherioll
4               None
             ...    
112585          None
112586          None
112587    RainMakers
112588          None
112589          None
Name: brand, Length: 112590, dtype: object

In [47]:
df["brand"].value_counts()

brand
Bath & Body Works    380
WILLBOND             318
Hairro               290
S-noilite            286
Generic              254
                    ... 
Stuvz                  1
Hooin                  1
SIYAN                  1
Flameer                1
KAIQIKAIXI             1
Name: count, Length: 22961, dtype: int64

In [49]:
df["brand"].nunique()

22961

In [51]:
df

title  average_rating  \
0       Howard LC0008 Leather Conditioner, 8-Ounce (4-...             4.8   
1       Yes to Tomatoes Detoxifying Charcoal Cleanser ...             4.5   
2        Eye Patch Black Adult with Tie Band (6 Per Pack)             4.4   
3       Tattoo Eyebrow Stickers, Waterproof Eyebrow, 4...             3.1   
4       Precision Plunger Bars for Cartridge Grips – 9...             4.3   
...                                                   ...             ...   
112585  TOPREETY 24"120gr 3/4 Full Head clip in hair e...             5.0   
112586  Pets Playmate Pet Grooming Glove,Gentle Deshed...             5.0   
112587  [10Pack] Makeup Brushes Set Cosmetics Tools Ki...             4.3   
112588  Xcoser Pretty Party Anna Wig Hair Tails Hair S...             4.0   
112589  DVIO Men's Voyage Perfume, Spicy woody fragran...             1.8   

        rating_number                                           features  \
0                  10                                                 []   
1                   3                                                 []   
2                  26                                                 []   
3                 102                                                 []   
4                   7  [Material: 304 Stainless Steel; Brass tip, Len...   
...               ...                                                ...   
112585              3                                                 []   
112586              1                                                 []   
112587              5                                                 []   
112588              9                                                 []   
112589              2  [Features: Spicy woody notes, pure sandalwood ...   

                                              description  \
0                                                      []   
1                                                      []   
2                                                      []   
3                                                      []   
4       [The Precision Plunger Bars are designed to wo...   
...                                                   ...   
112585                                                 []   
112586                                                 []   
112587                                                 []   
112588                                                 []   
112589                                                 []   

                                                   images  \
0       [{'thumb': 'https://m.media-amazon.com/images/...   
1       [{'thumb': 'https://m.media-amazon.com/images/...   
2       [{'thumb': 'https://m.media-amazon.com/images/...   
3       [{'thumb': 'https://m.media-amazon.com/images/...   
4       [{'thumb': 'https://m.media-amazon.com/images/...   
...                                                   ...   
112585  [{'thumb': 'https://m.media-amazon.com/images/...   
112586  [{'thumb': 'https://m.media-amazon.com/images/...   
112587  [{'thumb': 'https://m.media-amazon.com/images/...   
112588  [{'thumb': 'https://m.media-amazon.com/images/...   
112589  [{'thumb': 'https://m.media-amazon.com/images/...   

                         store  \
0              Howard Products   
1                       Yes To   
2       Levine Health Products   
3                     Cherioll   
4                    Precision   
...                        ...   
112585                TOPREETY   
112586           Pets Playmate   
112587              RainMakers   
112588                  Xcoser   
112589                    DVIO   

                                                  details       brand  
0       {'Package Dimensions': '7.1 x 5.5 x 3 inches; ...        None  
1       {'Item Form': 'Powder', 'Skin Type': 'Acne Pro...      Yes To  
2              {'Manufacturer': 'Levine Health Products'}        None  
3       {'Brand': 'Cherioll', 'Item Form': 'Powder'

So, if we count only "brands" that appear in the "details" field, we have less brands than stores, so we have 22961 brands. 

## Which brands have the highest ratings?

If we consider having highest ratining as having hte highest average_rating, then we must collect all the brands that have average_rating = 5.0

We have two scenarios: if stores are counted as brands OR if only those rows that contains a brand in brand column are counted as brands

In [147]:
#first case scenario: if stores are counted as brands 

In [89]:
df.groupby("store")["average_rating"].mean().sort_values(ascending=False) #we want the average, so we use the mean 
#with the groupby 

store
ESP                      5.0
kaBOOM! / Nelvana        5.0
TKB Trading, LLC         5.0
Find All                 5.0
FinchBerry               5.0
                        ... 
Aokin                    1.0
Herchuse                 1.0
oufenli                  1.0
Netual                   1.0
First Class Beauty Co    1.0
Name: average_rating, Length: 30761, dtype: float64

In [91]:
df_rate = df.groupby("store")["average_rating"].mean().sort_values(ascending=False).reset_index(name='mean_rating')
#groupby to see the average score given to each store combined 
df_rate.head()

store  mean_rating
0                ESP          5.0
1  kaBOOM! / Nelvana          5.0
2   TKB Trading, LLC          5.0
3           Find All          5.0
4         FinchBerry          5.0

In [93]:
df_rate5 = df_rate[df_rate['mean_rating'] == 5.0] #only take the ones with the highest mean_rating of 5.0
df_rate5

store  mean_rating
0                   ESP          5.0
1     kaBOOM! / Nelvana          5.0
2      TKB Trading, LLC          5.0
3              Find All          5.0
4            FinchBerry          5.0
...                 ...          ...
3058               MTJJ          5.0
3059             InKach          5.0
3060        Alpha Tools          5.0
3061         A.T. Cross          5.0
3062            Ingenue          5.0

[3063 rows x 2 columns]

In [222]:
#pandas to list - source https://www.geeksforgeeks.org/how-to-convert-pandas-column-to-list/
#convert store column into a list to have the output of all the names of the stores that have the 5.0 rating, which is the highest one. 

lista  = df_rate5['store'].values.tolist()
print(lista)
#this list contains all stores names who have received 5.0 average rating 

['ESP', 'kaBOOM! / Nelvana', 'TKB Trading, LLC', 'Find All', 'FinchBerry', 'Organic Age Defense', 'Fijian Noni Soap', 'Organic Colour Systems', 'Organic Complexion', 'Fiddleestickks', 'Ficcare', 'BURBERRY', 'Fibevon', 'Feyuan', 'MLB', 'kameleon t-shirts', 'Organic Health', 'Organic Island', 'k-SHION', 'Ferskov', 'juvia', 'justneem', 'Ferrari', 'Organic Ocean', 'Organic Traditions', 'Orglamix', 'Oriamour', 'Fekey&JF', 'Xawy', 'Xanthys & CoCLOL', 'Optimum Advanced', 'FirsTime', 'Optimum MadayFormula', 'Optique De Lobo', 'Special-Shop', 'Trend StarterNUATE', 'XStar', 'MARUKKAN', 'Five Star Fragrances', 'XUYI', 'Fit & Fresh', 'BROSWJUN', 'Orajel', 'BRS', 'Fireball Fragrances', 'BTBGYP', 'Fire Beauty', 'Firbon', 'FingerMates, Inc', "Finger's", 'kaniers', 'Finger Angel', 'BSN', 'Orbitol', 'Ordenado', 'FineVine', 'FinePure', 'FineMusez', 'Xinblue baby', 'Oriental Trading Company', 'Orientica', 'Baisida', 'Badan Body', 'Faith Hill True', 'Fairytale Bridal Tiara', 'Faconnable', 'Y. S. PARK', 'F

In [208]:
#second case scenario: if we take the brand column created 

In [53]:
df.head()

title  average_rating  \
0  Howard LC0008 Leather Conditioner, 8-Ounce (4-...             4.8   
1  Yes to Tomatoes Detoxifying Charcoal Cleanser ...             4.5   
2   Eye Patch Black Adult with Tie Band (6 Per Pack)             4.4   
3  Tattoo Eyebrow Stickers, Waterproof Eyebrow, 4...             3.1   
4  Precision Plunger Bars for Cartridge Grips – 9...             4.3   

   rating_number                                           features  \
0             10                                                 []   
1              3                                                 []   
2             26                                                 []   
3            102                                                 []   
4              7  [Material: 304 Stainless Steel; Brass tip, Len...   

                                         description  \
0                                                 []   
1                                                 []   
2                                                 []   
3                                                 []   
4  [The Precision Plunger Bars are designed to wo...   

                                              images                   store  \
0  [{'thumb': 'https://m.media-amazon.com/images/...         Howard Products   
1  [{'thumb': 'https://m.media-amazon.com/images/...                  Yes To   
2  [{'thumb': 'https://m.media-amazon.com/images/...  Levine Health Products   
3  [{'thumb': 'https://m.media-amazon.com/images/...                Cherioll   
4  [{'thumb': 'https://m.media-amazon.com/images/...               Precision   

                                             details     brand  
0  {'Package Dimensions': '7.1 x 5.5 x 3 inches; ...      None  
1  {'Item Form': 'Powder', 'Skin Type': 'Acne Pro...    Yes To  
2         {'Manufacturer': 'Levine Health Products'}      None  
3  {'Brand': 'Cherioll', 'Item Form': 'Powder', '...  Cherioll  
4                            {'UPC': '644287689178'}      None

In [71]:
df_br = df.groupby("brand")["average_rating"].mean().sort_values(ascending=False).reset_index(name='mean_rating')
#groupby to see the average score given to each store combined 

In [73]:
df_br

brand  mean_rating
0             Bellisima          5.0
1             JELLY PET          5.0
2      Aurora Cosmetics          5.0
3             Rucci Inc          5.0
4                MEIUGV          5.0
...                 ...          ...
22956          Roalschr          1.0
22957            target          1.0
22958         Baby Hair          1.0
22959           Fujiuia          1.0
22960            DIVAIN          1.0

[22961 rows x 2 columns]

In [75]:
df_br5 = df_br[df_br['mean_rating'] == 5.0]  #only take the ones with the highest mean_rating of 5.0

#list conversion
lista_brand  = df_br5['brand'].values.tolist()
print(lista_brand)
#list of all brands with average rating of 5.0

['Bellisima', 'JELLY PET', 'Aurora Cosmetics', 'Rucci Inc', 'MEIUGV', 'Effadaine', 'JEHEHUBO', 'Californiamicroneedle', 'California Mango', 'Aurora Active', 'Effol', 'VBLab', 'Authentic Naturals', 'Calico Skin Care', 'SILK OIL OF MOROCCO MadayFormula', 'Ktyssp Nail Art', 'bonballoon', 'Calolea', 'NewLine', 'Aunt Jackies Kids', 'Ferskov', 'bulk buys', 'Y Et Beauté', 'Can Solivera', 'Tokyo Design', 'Campion', 'bows for belles', 'Nicexx', 'PAWOAH', 'Kure Bazaar', 'born to run', 'Cameo Collection', 'PDC Brands', 'Eduard Vogt', 'Cala Products', 'Ava Haircare', 'Egnaro', 'Avon Clean', 'Sunee Hair', 'bebe prince', 'Rue21 Rue 21', 'Kresto', 'Shield-Safety', 'CYEVA', 'Marycrafts horn hair forks', 'Avalanche', 'Pruvit', 'Fekey&JF', 'TTI Blakemore', 'bebe', "Eia's Corner", 'YANAGIYA', 'AvicerePersephone', 'KritterKlips', 'CZgem', 'NiceOne', 'Krud Kutter', 'Rue21', 'Aveeno Baby', 'Cafetec', 'JH Foveam', 'Rudmox', 'VCOSTORE', 'Shiaon', 'Golden Lady', 'PEIPEIER', 'Pudus', 'SILK OIL OF MOROCCO', 'Puk

In [83]:
df_br5['brand']

0              Bellisima
1              JELLY PET
2       Aurora Cosmetics
3              Rucci Inc
4                 MEIUGV
              ...       
1879             Woochie
1880            On Trend
1881          Stay-Matte
1882               EBHDS
1883       karma organic
Name: brand, Length: 1884, dtype: object

## Which products have the highest average rating? List the top 50. Use products with number of rating greater than 1000.

To define product: the title will be used as the indicator of a product.

we have to groupby title and average_rating

In [65]:
#better understanding of groupby - source: https://data36.com/pandas-tutorial-2-aggregation-and-grouping/

In [219]:
dfpd = df[df["rating_number"] > 1000] #select only the products with number of rating greater than 1000.

In [253]:
dfpd.sort_values('average_rating', ascending=False).head()

title  average_rating  \
63448  100 Letterpress Note Cards and Self Seal Envel...             4.9   
33374              Corn Huskers Lotion 7 oz ( Pack of 4)             4.9   
95465  Graham Sanek Neck Strips- 2.5 x 17.5-12 pks w/...             4.8   
76279             Tigi Bed Head Manipulator 2oz (3 PACK)             4.8   
93953  1" Metallic Red, Silver & Blue Stars Shape Foi...             4.8   

       rating_number                                           features  \
63448           2322  [SELF SEAL ENVELOPES: Perfect for bulk mailing...   
33374           1027  [Moisturizes and softens dry skin, Soothes irr...   
95465           1662  [Made of a soft, highly absorbent tissue, Neck...   
76279           1260                                             [tigi]   
93953          12203                                                 []   

                                             description  \
63448  [Luxury Note Cards at an Affordable Price, Sho...   
33374  [4 Bottles of CORN HUSKERS LOTION 7 oz   OIL F...   
95465          [Graham Sanek Neck Strips 720 Strips/ pk]   
76279                                                 []   
93953                                          [ASIN020]   

                                                  images                store  \
63448  [{'thumb': 'https://m.media-amazon.com/images/...     Opie's Paper Co.   
33374  [{'thumb': 'https://m.media-amazon.com/images/...    Johnson & Johnson   
95465  [{'thumb': 'https://m.media-amazon.com/images/...  Graham Professional   
76279  [{'thumb': 'https://m.media-amazon.com/images/...                 TIGI   
93953  [{'thumb': 'https://m.media-amazon.com/images/...      Garage Sale Pup   

                                                 details  len_title  
63448  {'Is Discontinued By Manufacturer': 'No', 'Pro...         51  
33374  {'Brand': 'Johnson & Johnson', 'Item Form': 'L...         51  
95465  {'Is Discontinued By Manufacturer': 'No', 'Pac...         51  
76279  {'Product Benefits': 'Frizz Control', 'Hair Ty...         51  
93953  {'Color': 'Asin020', 'Size': 'ASIN020', 'Brand...         51

In [227]:
df_prod = dfpd.groupby("title")["average_rating"].mean().sort_values(ascending=False).reset_index(name='mean_rating')
df_prod.head(50) #the top 50
#df_prod.iloc[:50]

title  mean_rating
0               Corn Huskers Lotion 7 oz ( Pack of 4)          4.9
1   100 Letterpress Note Cards and Self Seal Envel...          4.9
2   Pack of 2100 Colorful Dog Paw Print Stickers, ...          4.8
3               Broadway Vita-Lip Argan Oil Lip Gloss          4.8
4   3/4" Bright Red Round Color Coding Circle Dot ...          4.8
5                          Segbeauty empty bottle 290          4.8
6   Oral-B Pro-Health Comfort Plus Mint Floss, Min...          4.8
7   Secret Original Shower Fresh Scent Women's Inv...          4.8
8   4.5 Inch Hair Bows For Girls Grosgrain Ribbon ...          4.8
9   L7 MANNEQUIN Realistic Mannequin Head with Sho...          4.8
10             Tigi Bed Head Manipulator 2oz (3 PACK)          4.8
11  Pack of 1008 1" Round Color Coding Circle Dot ...          4.8
12          Tigi Bedhead Hard Head Hairspray (6 Pack)          4.8
13  Pack of 1600 3/8 Inch Round Neon Color Coding ...          4.8
14  Aquaphor Healing Ointment - Moisturizing Skin ...          4.8
15  Garage Sale Pup Sonic bark deterrents 1.5" Inc...          4.8
16  Pack of 4200 3/4" Round Color Coding Circle Do...          4.8
17  Graham Sanek Neck Strips- 2.5 x 17.5-12 pks w/...          4.8
18  Balneol Hygienic Cleansing Lotion, 3 oz. (Pack...          4.8
19  Red Heart Shaped Sticker Labels, 1 1/4" Diamet...          4.8
20  Vivaplex, 12, Clear, 2 oz Glass Bottles, with ...          4.8
21  Hello Natural Apple Flavor Kids Fluoride Free ...          4.8
22  Janeke Superbrush Brush (Green/Purple - 86SP22...          4.8
23  Hibiclens Antiseptic & Antimicrobial Skin Clea...          4.8
24           Jack Black MP 10 Nourishing Oil, 2 Fl Oz          4.8
25  Picoway 20 Pack Mouse Ears Solid Black and Red...          4.8
26  Hibiclens Surgical Scrub 32 oz. Bottle (#57532...          4.8
27  Self Adhesive Tape - Yuelong 24Pack 2” x 5 Yar...          4.8
28  Gauge Gear Ear Stretching Balm | 10 ml Jar | P...          4.8
29    Purpose Gentle Cleansing Wash, 6 oz (Pack of 2)          4.8
30  1" Metallic Red, Silver & Blue Stars Shape Foi...          4.8
31  NEWTGAN 20 PCS Mouse Ears for Birthday Party T...          4.8
32  1.5" Inch Round Kraft, Baked with Love, Sticke...          4.8
33  Clinique Dramatically Different Moisturizing L...          4.8
34  Cerave Moisturizing Cream With Pump For Normal...          4.8
35  1.5" Heart Shape Kraft Paper Thank You Adhesiv...          4.8
36  1.25 Inch Round Thank You Labels with Heart, K...          4.8
37  1 Inch Round Thank You Sticker Labels with Red...          4.8
38  1 Inch Round Thank You Sticker Labels in Scrip...          4.8
39  2-1/4" x 1-1/4" Chalkboard Black Paper Label S...          4.8
40  Soft & Dri, Aerosol, Antiperspirant Deodorant,...          4.8
41  1 Inch Round Thank You Sticker Labels in Scrip...          4.8
42  Rohto SKIN AQUA UV Super Moisture Gel 110g SPF...          4.7
43        Hairspray (Two-Disc Shake & Shimmy Edition)          4.7
44  Hair Weave Needle and Thread Kit, 2 Black Weav...          4.7
45  SAUCE BEAUTY Guacamole Whip Hair Mask - Deep C...          4.7
46  Rose Quartz Jade Roller and Gua Sha Massager k...          4.7
47  Bath & Body Works, Signature Collection Body L...          4.7
48  Helion Communications - Chakra Centers Referen...          4.7
49  Happy Birthday Box for Women - Unique Birthday...          4.7

## Do these top-rated products share common characteristics?

They all have long detailed titles, with very specific product characteristics. Some have repeated titles with slightly different variations.

They all have details:  'Product Dimensions', 'Manufacturer', and contain the extra information of 'Is Discontinued By Manufacturer'. Some contain more detailed and specific information such as 'Material Type Free', 'Use for', 'Hair Type', which changes accordingly to the purpose and function of the product. I believe that the beauty products contain more information than general use products. 

In [74]:
lis  = df_prod['title'].values.tolist()

In [76]:
for i in range(10):
  print(lis[i])

Corn Huskers Lotion 7 oz ( Pack of 4)
100 Letterpress Note Cards and Self Seal Envelopes - Perfect for Clients, Friends, Co-Workers - Blank Inside for Your Message - Opie’s Paper Company
Pack of 2100 Colorful Dog Paw Print Stickers, 3/4 inch, 8 Bright Neon Colors, Great for Teachers, Classrooms & Veterinarians!
Broadway Vita-Lip Argan Oil Lip Gloss
3/4" Bright Red Round Color Coding Circle Dot Labels on a Roll, Semi-Gloss, 1000 Stickers.75 inch Diameter.
Segbeauty empty bottle 290
Oral-B Pro-Health Comfort Plus Mint Floss, Mint,6 Count (Pack of 1)
Secret Original Shower Fresh Scent Women's Invisible Solid pH Balanced Antiperspirant and Deodorant Twin Pack, 5.2 Ounce
4.5 Inch Hair Bows For Girls Grosgrain Ribbon Boutique Bow Clips Teens Toddlers Kids Set Of 30 (15 Colors x 2)
L7 MANNEQUIN Realistic Mannequin Head with Shoulders Plastic mannequin Heads for Wigs Earrings Hat Sunglassess Display


In [249]:
df.iloc[33374].details

{'Brand': 'Johnson & Johnson',
 'Item Form': 'Lotion',
 'Unit Count': '28 Fl Oz',
 'Number of Items': '4',
 'Use for': 'Whole Body',
 'Is Discontinued By Manufacturer': 'No',
 'Product Dimensions': '2.57 x 1.25 x 6.75 inches; 8 Ounces',
 'Item model number': 'Hand Lotion',
 'UPC': '885830512752 312547481274',
 'Manufacturer': 'Johnson & Johnson'}

In [251]:
df.iloc[63448].details

{'Is Discontinued By Manufacturer': 'No',
 'Product Dimensions': '11.5 x 4.2 x 5.2 inches; 3.24 Pounds',
 'UPC': '648499871505',
 'Manufacturer': "Opie's Paper Co."}

In [255]:
df.iloc[76279].details

{'Product Benefits': 'Frizz Control',
 'Hair Type': 'All',
 'Material Type Free': 'Benzyl Alcohol',
 'Scent': 'tropical',
 'Liquid Volume': '2 Fluid Ounces',
 'Is Discontinued By Manufacturer': 'No',
 'Product Dimensions': '2 x 2 x 2 inches; 8.64 Ounces',
 'Item model number': '942110',
 'UPC': '787734557924 763985661535 632181674628',
 'Manufacturer': 'Tigi'}

## Display images and print details of the top 10 highest-rated products.

In [79]:
#merge both df_prod and df together to get the previous columns back
ndf = pd.merge(df_prod, df, how='left', on='title')
ndf = ndf.head(10)
ndf

title  mean_rating  \
0              Corn Huskers Lotion 7 oz ( Pack of 4)          4.9   
1  100 Letterpress Note Cards and Self Seal Envel...          4.9   
2  Pack of 2100 Colorful Dog Paw Print Stickers, ...          4.8   
3              Broadway Vita-Lip Argan Oil Lip Gloss          4.8   
4  3/4" Bright Red Round Color Coding Circle Dot ...          4.8   
5                         Segbeauty empty bottle 290          4.8   
6  Oral-B Pro-Health Comfort Plus Mint Floss, Min...          4.8   
7  Secret Original Shower Fresh Scent Women's Inv...          4.8   
8  4.5 Inch Hair Bows For Girls Grosgrain Ribbon ...          4.8   
9  L7 MANNEQUIN Realistic Mannequin Head with Sho...          4.8   

   average_rating  rating_number  \
0             4.9           1027   
1             4.9           2322   
2             4.8          10775   
3             4.8           1704   
4             4.8          12205   
5             4.8           4920   
6             4.8           2109   
7             4.8           1010   
8             4.8           1085   
9             4.8           1028   

                                            features  \
0  [Moisturizes and softens dry skin, Soothes irr...   
1  [SELF SEAL ENVELOPES: Perfect for bulk mailing...   
2                                                 []   
3                                                 []   
4                                                 []   
5                                                 []   
6  [Contains 6 packs of Oral-B Glide Pro-Health C...   
7  [You will receive a Pack of 2 Secret Original ...   
8                                                 []   
9  [This mannequin head Suitable for different st...   

                                         description  \
0  [4 Bottles of CORN HUSKERS LOTION 7 oz   OIL F...   
1  [Luxury Note Cards at an Affordable Price, Sho...   
2                                                 []   
3                                                 []   
4                                                 []   
5                                                 []   
6  [Oral-B Glide Pro-Heath Comfort Plus Floss is ...   
7  [Stay true with Secret. You can always be conf...   
8                                                 []   
9                                                 []   

                                              images              store  \
0  [{'thumb': 'https://m.media-amazon.com/images/...  Johnson & Johnson   
1  [{'thumb': 'https://m.media-amazon.com/images/...   Opie's Paper Co.   
2  [{'thumb': 'https://m.media-amazon.com/images/...    Garage Sale Pup   
3  [{'thumb': 'https://m.media-amazon.com/images/...           Broadway   
4  [{'thumb': 'https://m.media-amazon.com/images/...    Garage Sale Pup   
5  [{'thumb': 'https://m.media-amazon.com/images/...             TruuMe   
6  [{'thumb': 'https://m.media-amazon.com/images/...              GLIDE   
7  [{'thumb': 'https://m.media-amazon.com/images/...             Secret   
8  [{'thumb': 'https://m.media-amazon.com/images/...              Manwe   
9  [{'thumb': 'https://m.media-amazon.com/images/...       L7 MANNEQUIN   

                                             details  
0  {'Brand': 'Johnson & Johnson', 'Item Form': 'L...  
1  {'Is Discontinued By Manufacturer': 'No', 'Pro...  
2  {'Color': 'Pcs-348', 'Size': 'Pcs-348', 'Brand...  
3  {'Brand': 'Broadway', 'Skin Type': 'All', 'Ite...  
4  {'Color': 'Pcs-361', 'Size': 'pcs-361', 'Brand...  
5  {'Brand': 'Segbeauty', 'Color': 'Clear', 'Mate...  
6  {'Brand': 'GLIDE', 'Dental Floss Type': 'Denta...  
7  {'Package Dimensions': '8.43 x 6.89 x 4.41 inc...  
8  {'Brand': 'Manwe', 'Hair Type': 'All', 'Age Ra...  
9  {'Brand': 'L7 MANNEQUIN', 'Color': 'Black', 'S...

In [81]:
#using join df_prod.set_index('title').join(df.set_index('title'))

In [83]:
#using join df_prod.join(df.set_index('title'), on='title')

In [85]:
#display image link 
for index,row in ndf.iterrows():
    print(ndf.iloc[index].images[0]['large'])

https://m.media-amazon.com/images/I/51+2zxwHeVL.jpg
https://m.media-amazon.com/images/I/419oi-H1DeL.jpg
https://m.media-amazon.com/images/I/61T3tYlsruL.jpg
https://m.media-amazon.com/images/I/51QBvJ6B8ZL.jpg
https://m.media-amazon.com/images/I/31UKMleR14L.jpg
https://m.media-amazon.com/images/I/31oTJurWykL.jpg
https://m.media-amazon.com/images/I/51Y7vkuPe-L.jpg
https://m.media-amazon.com/images/I/41bs9ujDABL.jpg
https://m.media-amazon.com/images/I/51yCvC34+XL.jpg
https://m.media-amazon.com/images/I/31cxNDwjJBL.jpg


In [87]:
#Details

In [452]:
detalhes  = ndf['details'].values.tolist()
print(detalhes, '\n')

[{'Brand': 'Johnson & Johnson', 'Item Form': 'Lotion', 'Unit Count': '28 Fl Oz', 'Number of Items': '4', 'Use for': 'Whole Body', 'Is Discontinued By Manufacturer': 'No', 'Product Dimensions': '2.57 x 1.25 x 6.75 inches; 8 Ounces', 'Item model number': 'Hand Lotion', 'UPC': '885830512752 312547481274', 'Manufacturer': 'Johnson & Johnson'}, {'Is Discontinued By Manufacturer': 'No', 'Product Dimensions': '11.5 x 4.2 x 5.2 inches; 3.24 Pounds', 'UPC': '648499871505', 'Manufacturer': "Opie's Paper Co."}, {'Color': 'Pcs-348', 'Size': 'Pcs-348', 'Brand': 'Garage Sale Pup', 'Shape': 'Round', 'Number of Labels': '2100', 'Is Discontinued By Manufacturer': 'No', 'Product Dimensions': '11 x 0.13 x 8.5 inches; 5.92 Ounces', 'Manufacturer': 'Garage Sale Pup'}, {'Brand': 'Broadway', 'Skin Type': 'All', 'Item Form': 'Oil', 'Finish Type': 'Glossy', 'Product Benefits': 'Soothing, Moisturizing', 'Package Dimensions': '5.55 x 4.29 x 0.71 inches; 3.84 Ounces', 'UPC': '731509807011'}, {'Color': 'Pcs-361', 

In [448]:
df['details']

0         {'Package Dimensions': '7.1 x 5.5 x 3 inches; ...
1         {'Item Form': 'Powder', 'Skin Type': 'Acne Pro...
2                {'Manufacturer': 'Levine Health Products'}
3         {'Brand': 'Cherioll', 'Item Form': 'Powder', '...
4                                   {'UPC': '644287689178'}
                                ...                        
112585    {'Is Discontinued By Manufacturer': 'No', 'Pac...
112586    {'Is Discontinued By Manufacturer': 'No', 'Pac...
112587    {'Brand': 'RainMakers', 'Recommended Uses For ...
112588    {'Is Discontinued By Manufacturer': 'No', 'Pac...
112589      {'UPC': '736238729337', 'Manufacturer': 'DVIO'}
Name: details, Length: 112590, dtype: object

In [538]:
for index,row in ndf.iterrows():
    print(ndf.details.iloc[index],'\n')

{'Brand': 'Johnson & Johnson', 'Item Form': 'Lotion', 'Unit Count': '28 Fl Oz', 'Number of Items': '4', 'Use for': 'Whole Body', 'Is Discontinued By Manufacturer': 'No', 'Product Dimensions': '2.57 x 1.25 x 6.75 inches; 8 Ounces', 'Item model number': 'Hand Lotion', 'UPC': '885830512752 312547481274', 'Manufacturer': 'Johnson & Johnson'} 

{'Is Discontinued By Manufacturer': 'No', 'Product Dimensions': '11.5 x 4.2 x 5.2 inches; 3.24 Pounds', 'UPC': '648499871505', 'Manufacturer': "Opie's Paper Co."} 

{'Color': 'Pcs-348', 'Size': 'Pcs-348', 'Brand': 'Garage Sale Pup', 'Shape': 'Round', 'Number of Labels': '2100', 'Is Discontinued By Manufacturer': 'No', 'Product Dimensions': '11 x 0.13 x 8.5 inches; 5.92 Ounces', 'Manufacturer': 'Garage Sale Pup'} 

{'Brand': 'Broadway', 'Skin Type': 'All', 'Item Form': 'Oil', 'Finish Type': 'Glossy', 'Product Benefits': 'Soothing, Moisturizing', 'Package Dimensions': '5.55 x 4.29 x 0.71 inches; 3.84 Ounces', 'UPC': '731509807011'} 

{'Color': 'Pcs-361

## What will be a good marketing strategy when your company launches its first product?

I think that it is imporatnt to have the image in high_resolution as well, as all the others, as the most complete the information, the better it will be for the consumer and whoever more sees the data.
Add details, with as much information as possible and by also adding features and description it might help increase sales, as we can see that some of the top 10 sellers have all those columns filled. 
Maybe it will be goos to have the same product, from the same brand, selling at different stores, to increase their market outreach. 

---

## 1. Can a product not have a store defined (NaN value) and still have good average review (higher than 4.5) and rating_number over 1000?

In [10]:
df.isnull().sum()

title                12
average_rating        0
rating_number         0
features              0
description           0
images                0
store             11344
details               0
dtype: int64

In [71]:
dfnostore = df[df['store'].isnull()] #isna() ok too
#store all the NaN stores in this new df 

In [83]:
dfnostore.head(1)

title  average_rating  \
12  Brand New Headrang Face line Contour V-line Ma...             3.3   

    rating_number features description  \
12              4       []          []   

                                               images store  \
12  [{'thumb': 'https://m.media-amazon.com/images/...   NaN   

                                              details  
12  {'Is Discontinued By Manufacturer': 'No', 'Pac...

In [86]:
dfnostore4 = dfnostore[dfnostore['average_rating'] >= 4.5]
len(dfnostore4)

2924

In [115]:
dfnostore44 = dfnostore4[dfnostore4['rating_number'] >= 1000]
len(dfnostore44)

15

In [117]:
dfnostore44['average_rating'].value_counts()

average_rating
4.6    7
4.7    4
4.5    3
4.8    1
Name: count, dtype: int64

In [119]:
df_dep = dfnostore44.groupby("average_rating")["rating_number"].mean().reset_index(name='total')
df_dep

average_rating        total
0             4.5  1860.000000
1             4.6  2221.428571
2             4.7  2219.000000
3             4.8  1040.000000

there are only 15 stores out of the 

## 2. How many products without description have been sold?

null descriptions are []

In [170]:
df.isnull().sum()

title                12
average_rating        0
rating_number         0
features              0
description           0
images                0
store             11344
details               0
dtype: int64

In [172]:
dfnodesc = df.dropna()

In [174]:
dfnodesc = dfnodesc[dfnodesc['description'] == '[]']
print(len(dfnodesc))
dfnodesc.head(3)

82617


title  average_rating  \
0  Howard LC0008 Leather Conditioner, 8-Ounce (4-...             4.8   
1  Yes to Tomatoes Detoxifying Charcoal Cleanser ...             4.5   
2   Eye Patch Black Adult with Tie Band (6 Per Pack)             4.4   

   rating_number features description  \
0             10       []          []   
1              3       []          []   
2             26       []          []   

                                              images                   store  \
0  [{'thumb': 'https://m.media-amazon.com/images/...         Howard Products   
1  [{'thumb': 'https://m.media-amazon.com/images/...                  Yes To   
2  [{'thumb': 'https://m.media-amazon.com/images/...  Levine Health Products   

                                             details  
0  {'Package Dimensions': '7.1 x 5.5 x 3 inches; ...  
1  {'Item Form': 'Powder', 'Skin Type': 'Acne Pro...  
2         {'Manufacturer': 'Levine Health Products'}

82617 products without description have been sold.

## 3. How many products without details have been sold and gotten an average rating higher and equal than 4.5?

Assuming that empty values are {} 

There are 890 products

In [182]:
dfnodt = df.dropna() #drop all null rows 
dfnodt = dfnodt[dfnodt['average_rating'] >= 4.5]

In [184]:
dfnodt = dfnodt[dfnodt['details'] == '{}']
print(len(dfnodt))
dfnodt.head(3)

890


title  average_rating  \
36   Luxury Shimmer Pearl Loose Eyeshadow Powder Ey...             5.0   
467  Affiance Infiniti Intelligent Colour System Ul...             5.0   
531  Panthere De Cartier 2-piece Set 1.6 Oz Parfum ...             5.0   

     rating_number features description  \
36               1       []          []   
467              1       []          []   
531              1       []          []   

                                                images              store  \
36   [{'thumb': 'https://m.media-amazon.com/images/...            WindMax   
467  [{'thumb': 'https://m.media-amazon.com/images/...  Affinage Infiniti   
531  [{'thumb': 'https://m.media-amazon.com/images/...             ERTDDE   

    details  
36       {}  
467      {}  
531      {}

## 4. What are the most common Age Range in details?

Adult is the most common one, with 21117 ocurances 

In [341]:
def aging(x):
    if isinstance(x, Dict):
        for key, value in x.items():
            if key == 'Age Range (Description)':
                return value
    else:
        return None

In [343]:
df["age"] = df["details"].apply(aging) #apply function to df and create column brand 
df["age"]

0          None
1         Adult
2          None
3          None
4          None
          ...  
112585     None
112586     None
112587    Adult
112588     None
112589     None
Name: age, Length: 112590, dtype: object

In [345]:
df["age"].value_counts()

age
Adult            21117
Kid                522
Child              409
Baby               267
All Ages           109
                 ...  
Mayores de 12        1
14-99                1
12 Years & Up        1
Newborn,Baby         1
gel face wash        1
Name: count, Length: 72, dtype: int64

## 5. What is the average lenght of a title, and what is the shortest title and the longest title?

In [109]:
lendf = df.dropna()

In [216]:
lendf['length_title'] = lendf['title'].apply(len) #Get the length of each title

/var/folders/5q/byw5947x45z270nws3pnkw440000gn/T/ipykernel_68169/1014798506.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  lendf['length_title'] = lendf['title'].apply(len) #Get the length of each title


In [119]:
lendf.head(3)

title  average_rating  \
0  Howard LC0008 Leather Conditioner, 8-Ounce (4-...             4.8   
1  Yes to Tomatoes Detoxifying Charcoal Cleanser ...             4.5   
2   Eye Patch Black Adult with Tie Band (6 Per Pack)             4.4   

   rating_number features description  \
0             10       []          []   
1              3       []          []   
2             26       []          []   

                                              images                   store  \
0  [{'thumb': 'https://m.media-amazon.com/images/...         Howard Products   
1  [{'thumb': 'https://m.media-amazon.com/images/...                  Yes To   
2  [{'thumb': 'https://m.media-amazon.com/images/...  Levine Health Products   

                                             details  len_title  length  \
0  {'Package Dimensions': '7.1 x 5.5 x 3 inches; ...         51      51   
1  {'Item Form': 'Powder', 'Skin Type': 'Acne Pro...         51     143   
2         {'Manufacturer': 'Levine Health Products'}         51      48   

   length_title  
0            51  
1           143  
2            48

In [145]:
print('The average length of a title is', lendf["length_title"].mean())

The average length of a title is 115.51350310166345


In [137]:
print('The longest title has ',lendf["length_title"].max(),'characters')

The longest title has  1455 characters


In [139]:
print('The shorstes title has', lendf["length_title"].min(), 'characters')

The shorstes title has 1 characters


## 6. How many different products do the top 5 stores sold?

groupby store and title

Bath & Body Works has 477 unique products, Generic has 368, Hairro has 378, S-noilite has 340 and WILLBOND has 386

So, it seems that no store has sold multiple times the same item, they are all different. 

In [99]:
df_5 = df.dropna()

title             0
average_rating    0
rating_number     0
features          0
description       0
images            0
store             0
details           0
dtype: int64

In [193]:
df_5['store'].value_counts()

store
Bath & Body Works    481
WILLBOND             388
Hairro               381
Generic              369
S-noilite            340
                    ... 
ASSOS Cycling          1
LA Bodies              1
Lingon                 1
STAUB                  1
DVIO                   1
Name: count, Length: 30759, dtype: int64

In [157]:
dftop = df_5['store'].value_counts().reset_index().head() #the top 5 companies that appears the most 
dftop

store  count
0  Bath & Body Works    481
1           WILLBOND    388
2             Hairro    381
3            Generic    369
4          S-noilite    340

In [161]:
#make it to a list the store names
lis = dftop['store'].values.tolist()
lis

['Bath & Body Works', 'WILLBOND', 'Hairro', 'Generic', 'S-noilite']

In [173]:
#count how many different products they sell
# use https://discovery.cs.illinois.edu/guides/DataFrame-Row-Selection/dataframe-isin-selection/#:~:text=isin%20function%20is%20commonly%20used,they%20are%20in%20another%20DataFrame.
# to understand .isin function
# and https://www.w3schools.com/python/pandas/ref_df_isin.asp#:~:text=The%20isin()%20method%20checks,the%20specified%20values%2C%20otherwise%20False%20.
#for more examples

In [189]:
# def getall(i):
#     if i in lis:
#         return i

# d5 = df_5[df_5['store'].apply(getall)]
#-- this outputs an error, therefore I had to search another alternative which was isin()

In [219]:
dftop5full = df_5[df_5['store'].isin(lis)]
dftop5full.head(2)

title  average_rating  \
132  Bath & Body Works PocketBac Hand Gel I'm Reall...             3.5   
133  Halloween Wizard Wig and Faux Beard Set Party ...             3.7   

     rating_number features description  \
132              6       []          []   
133             42       []          []   

                                                images              store  \
132  [{'thumb': 'https://m.media-amazon.com/images/...  Bath & Body Works   
133  [{'thumb': 'https://m.media-amazon.com/images/...           WILLBOND   

                                               details  
132                                                 {}  
133  {'Color': 'Gray', 'Material': 'Synthetic', 'Ha...

In [225]:
dftotal = dftop5full.groupby("store")["title"].nunique() #find all the unique titles in the df corresponding to each store
dftotal

store
Bath & Body Works    477
Generic              368
Hairro               378
S-noilite            340
WILLBOND             386
Name: title, dtype: int64

## 7. How many different countries we have in the dataset, based on stores? based on the top 1000 stores

I have found https://pypi.org/project/langdetect/ langdetect library, so if we see how many different languages we have in stores, we can try to estimate how many countries are in this dataset. This would be more meaningful instead of just seeing if it is an store with english name/characters or not.

In [315]:
! pip install langdetect

In [321]:
from langdetect import detect

In [ ]:
for i in 

In [343]:
nomes = df_5['store'].value_counts().reset_index().head(100)
nomes.head()

store  count
0  Bath & Body Works    481
1           WILLBOND    388
2             Hairro    381
3            Generic    369
4          S-noilite    340

In [349]:
nomes['store'].apply(detect).value_counts()

store
de    24
en    15
sw     9
fr     7
tl     6
id     6
af     4
so     4
fi     3
lt     3
vi     3
es     2
it     2
ca     2
da     2
ro     2
cy     2
et     1
pl     1
sv     1
tr     1
Name: count, dtype: int64

This is a really interesting and surprising finding as there seems to be 24 German (de) stores, and 15 English speaking stores, 9 Swahili stores, 7 French, and 6 Philipines stores (tl), and 6 Indonesian stores. Accoridng to langdetect language detector, based on https://en.wikipedia.org/wiki/List_of_ISO_639_language_codes#de abbreviations. 